src/gs-side-filter-row.c

```diff
@@ -0,0 +1,146 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gtk/gtk.h>
+
+#include "gs-side-filter-row.h"
+#include "gs-common.h"
+
+struct _GsSideFilterRow
+{
+	GtkListBoxRow	parent_instance;
+
+	GsCategory	*cat;
+	GsShellMode	 mode;
+	GtkWidget	*label;
+	GtkWidget	*image;
+	GtkWidget	*leftborder;
+};
+
+G_DEFINE_TYPE (GsSideFilterRow, gs_side_filter_row, GTK_TYPE_LIST_BOX_ROW)
+
+GsCategory *
+gs_side_filter_row_get_category (GsSideFilterRow *row)
+{
+	g_return_val_if_fail (GS_IS_SIDE_FILTER_ROW (row), NULL);
+
+	return row->cat;
+}
+
+void
+gs_side_filter_row_set_category (GsSideFilterRow *row, GsCategory *cat)
+{
+	GPtrArray *key_colors;
+
+	g_return_if_fail (GS_IS_SIDE_FILTER_ROW (row));
+	g_return_if_fail (GS_IS_CATEGORY (cat));
+
+	g_clear_object (&row->cat);
+	row->cat = g_object_ref (cat);
+
+	gtk_label_set_label (GTK_LABEL (row->label), gs_category_get_name (cat));
+	gtk_image_set_from_icon_name (GTK_IMAGE (row->image),
+				      gs_category_get_icon (cat),
+				      GTK_ICON_SIZE_MENU);
+
+	/* set custom CSS for the colored border */
+	key_colors = gs_category_get_key_colors (cat);
+	if (key_colors->len > 0) {
+		GdkRGBA *tmp = g_ptr_array_index (key_colors, 0);
+		g_autofree gchar *css = NULL;
+		g_autofree gchar *color = gdk_rgba_to_string (tmp);;
+		css = g_strdup_printf ("background-color: %s", color);
+		gs_utils_widget_set_css_simple (GTK_WIDGET (row->leftborder),
+						css);
+	}
+
+	gs_side_filter_row_set_mode (row, GS_SHELL_MODE_CATEGORY);
+}
+
+static void
+gs_side_filter_row_destroy (GtkWidget *widget)
+{
+	GsSideFilterRow *row = GS_SIDE_FILTER_ROW (widget);
+
+	g_clear_object (&row->cat);
+
+	GTK_WIDGET_CLASS (gs_side_filter_row_parent_class)->destroy (widget);
+}
+
+static void
+gs_side_filter_row_init (GsSideFilterRow *row)
+{
+	gtk_widget_set_has_window (GTK_WIDGET (row), FALSE);
+	gtk_widget_init_template (GTK_WIDGET (row));
+	row->mode = GS_SHELL_MODE_UNKNOWN;
+}
+
+static void
+gs_side_filter_row_class_init (GsSideFilterRowClass *klass)
+{
+	GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
+
+	widget_class->destroy = gs_side_filter_row_destroy;
+
+	gtk_widget_class_set_template_from_resource (widget_class,
+						     "/org/gnome/Software/"
+						     "gs-side-filter-row.ui");
+
+	gtk_widget_class_bind_template_child (widget_class, GsSideFilterRow,
+					      label);
+	gtk_widget_class_bind_template_child (widget_class, GsSideFilterRow,
+					      image);
+	gtk_widget_class_bind_template_child (widget_class, GsSideFilterRow,
+					      leftborder);
+}
+
+GtkWidget *
+gs_side_filter_row_new (GsCategory *cat)
+{
+	GsSideFilterRow *row;
+
+	row = g_object_new (GS_TYPE_SIDE_FILTER_ROW, NULL);
+	gs_side_filter_row_set_category (row, cat);
```

Comment:
I usually try to keep the _new() function as dumb as possible (i.e. only call `g_object_new()`), because that's what would be expected when using a class from bindings (not an issue in this case).
Still you could either have the caller call the setter, or make the category a property you also set in the call to `g_object_new()`.